coordinator/src/orderbook/mod.rs

```diff
@@ -30,120 +41,317 @@ pub mod websocket;
+        // process orderbook updates much faster in memory and postpone the expensive database
+        // transactions.
         let order = spawn_blocking({
             let mut conn = self.pool.clone().get()?;
             move || {
-                let order = match new_order {
-                    NewOrder::Market(o) => {
-                        orders::insert_market_order(&mut conn, o.clone(), order_reason)
-                    }
-                    NewOrder::Limit(o) => orders::insert_limit_order(&mut conn, o, order_reason),
-                }
-                .map_err(|e| anyhow!(e))
-                .context("Failed to insert new order into DB")?;
+                let order = orders::insert_market_order(&mut conn, new_order, order_reason)
+                    .map_err(|e| anyhow!(e))
+                    .context("Failed to insert new order into DB")?;
 
                 anyhow::Ok(order)
             }
         })
         .await??;
 
-        let message = match order.order_type {
-            OrderType::Market => None,
-            OrderType::Limit => {
-                match order.direction {
-                    commons::Direction::Short => self.shorts.add(order.clone()),
-                    commons::Direction::Long => self.longs.add(order.clone()),
-                }
-                Some(Message::NewOrder(order))
+        // find a match for the market order.
+        let matched_orders = match order.direction.opposite() {
+            commons::Direction::Short => self.shorts.match_order(order.quantity),
+            commons::Direction::Long => self.longs.match_order(order.quantity),
+        };
+
+        let fail_order = {
+            let order_id = order.id;
+            let pool = self.pool.clone();
+            move || {
+                let mut conn = pool.get()?;
+                orders::set_order_state(&mut conn, order_id, OrderState::Failed)?;
+                anyhow::Ok(())
             }
         };
 
-        Ok(message)
+        let matched_quantity: Decimal = matched_orders.iter().map(|o| o.quantity).sum();
+        if matched_quantity != order.quantity {
+            // not enough liquidity in the orderbook.
+            tracing::warn!(
+                trader_pubkey = %order.trader_id,
+                order_id = %order.id,
+                wanted = %order.quantity,
+                got = %matched_quantity,
+                "Couldn't match order due to insufficient liquidity in the orderbook."
+            );
+            match order.direction.opposite() {
+                commons::Direction::Short => self.shorts.rollback_matched_orders(matched_orders),
+                commons::Direction::Long => self.longs.rollback_matched_orders(matched_orders),
+            }
+
+            spawn_blocking(fail_order).await??;
+            bail!("No match found.");
+        }
+
+        // apply match to database.
+        let matches = spawn_blocking({
```

Comment:
🤔 This whole block scares me. It looks like we have already established a match using our in-memory orderbook, but now have to go to the database a bunch of times to get this done. I think we want to defer most of this work, and avoid having to load the `matching_fee` here.